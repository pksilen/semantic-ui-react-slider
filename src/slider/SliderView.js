/* eslint-disable react/jsx-props-no-spreading */
// @flow

import React, { useCallback } from 'react';
import type { Element } from 'react';
// $FlowFixMe
import PropTypes from 'prop-types';
// $FlowFixMe
import { Slider as ReactCompoundSlider, Handles, Rail, Tracks } from 'react-compound-slider';
// $FlowFixMe
import SliderHandleView from '../sliderhandle/SliderHandleView';
// $FlowFixMe
import SliderTrackView from '../slidertrack/SliderTrackView';
// $FlowFixMe
import type { SliderHandle } from '../sliderhandle/SliderHandle';

type SliderProps = $Exact<{
  className?: ?string,
  onSliderValuesChange: (number, number) => void,
  selectedMaxValue?: ?number,
  selectedMinValue?: ?number,
  sliderMaxValue: number,
  sliderMinValue: number
}>;

const SliderView = ({
  className,
  onSliderValuesChange,
  selectedMaxValue,
  selectedMinValue,
  sliderMaxValue,
  sliderMinValue
}: SliderProps): Element<any> => {
  const handleSliderUpdate = useCallback((values: number[]) => {
    if (values.length === 2) {
      onSliderValuesChange(values[0], values[1]);
    }
  }, []);

  const sliderStyle = {
    position: 'relative',
    height: '45px',
    marginLeft: '10px',
    marginRight: '10px',
    flexBasis: 'calc(100% - 20px)'
  };

  const sliderRailStyle = {
    backgroundColor: 'rgb(150, 150, 150)',
    borderRadius: '5px',
    height: '10px',
    marginTop: '18px',
    position: 'absolute',
    width: '100%'
  };

  return (
    <ReactCompoundSlider
      className={className}
      rootStyle={sliderStyle}
      domain={[sliderMinValue, sliderMaxValue]}
      step={1}
      mode={2}
      values={[selectedMinValue ?? sliderMinValue, selectedMaxValue ?? sliderMaxValue]}
      onUpdate={handleSliderUpdate}
    >
      <Rail>{({ getRailProps }: any) => <div style={sliderRailStyle} {...getRailProps()} />}</Rail>
      <Handles>
        {({ handles, getHandleProps }: any): Element<any> => (
          <div className="slider-handles">
            {handles.map((handle: SliderHandle) => (
              <SliderHandleView key={handle.id} handle={handle} getHandleProps={getHandleProps} />
            ))}
          </div>
        )}
      </Handles>
      <Tracks left={false} right={false}>
        {({ tracks, getTrackProps }: any): Element<any> => (
          <div className="slider-tracks">
            {tracks.map(({ id, source, target }: any) => (
              <SliderTrackView
                key={id}
                sourceHandle={source}
                targetHandle={target}
                getTrackProps={getTrackProps}
              />
            ))}
          </div>
        )}
      </Tracks>
    </ReactCompoundSlider>
  );
};

SliderView.propTypes = {
  className: PropTypes.string,
  onSliderValuesChange: PropTypes.func.isRequired,
  selectedMaxValue: PropTypes.number,
  selectedMinValue: PropTypes.number,
  sliderMaxValue: PropTypes.number.isRequired,
  sliderMinValue: PropTypes.number.isRequired
};

SliderView.defaultProps = {
  className: '',
  selectedMinValue: null,
  selectedMaxValue: null
};

export default SliderView;
