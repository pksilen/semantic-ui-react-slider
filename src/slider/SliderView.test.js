import React from 'react';
import { shallow as renderShallow } from 'enzyme';
import SliderView from './SliderView';

describe('render', () => {
  it('should render SliderView correctly', () => {
    // WHEN
    const sliderView = renderShallow(
      <SliderView sliderMinValue={0} sliderMaxValue={0} onSliderValuesChange={jest.fn()} />
    );

    expect(sliderView).toMatchSnapshot();
  });
});
