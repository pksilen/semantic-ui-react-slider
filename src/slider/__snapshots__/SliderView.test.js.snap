// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render SliderView correctly 1`] = `
<Slider
  className=""
  component="div"
  disabled={false}
  domain={
    Array [
      0,
      0,
    ]
  }
  flatten={false}
  mode={2}
  onChange={[Function]}
  onSlideEnd={[Function]}
  onSlideStart={[Function]}
  onUpdate={[Function]}
  reversed={false}
  rootProps={Object {}}
  rootStyle={
    Object {
      "flexBasis": "calc(100% - 20px)",
      "height": "45px",
      "marginLeft": "10px",
      "marginRight": "10px",
      "position": "relative",
    }
  }
  step={1}
  values={
    Array [
      0,
      0,
    ]
  }
  vertical={false}
  warnOnChanges={false}
>
  <Rail>
    <Component />
  </Rail>
  <Handles>
    <Component />
  </Handles>
  <Tracks
    left={false}
    right={false}
  >
    <Component />
  </Tracks>
</Slider>
`;
